runtime/scripts/results.js

```diff
@@ -0,0 +1,211 @@
+        //var ensure_decimal = math.ensure_decimal = function(n) { ? We need them to be like this to ensure they're accessible from elsewhere, maybe.
+        
+        /** Ensures a string will not cause issues within a csv due to commas, quotes, etc. 
+         * 
+         * @param {string} cell 
+         * @returns {string}
+         */
+        escape_csv_cell : function (cell) {
+            cell = cell + '';
+            if(cell.match(/[,"']/)) {
+                cell = '"'+cell.replace(/\\/g,'\\\\').replace(/"/g,'\\"')+'"';
+            }
+            return cell;
+        },
+        
+        /** Escapes each cell of a list of strings such that each will not cause issues within a csv
+         * 
+         * @param {Array.<string>} cells 
+         * @returns {Array.<string>} 
+         */
+        make_csv_row : function (cells) {
+            return cells.map(Numbas.results.escape_csv_cell).join(',');
+        },
+
+        /** Escapes each cell of a two-dimensional array of strings such that each will not cause issues within a csv
+         * 
+         * @param {Array.<Array.<string>>} rows 
+         * @returns  {Array.<Array.<string>>} 
+         */
+        make_csv : function (rows) {
+            return rows.map(Numbas.results.make_csv_row).join('\n');
+        },
+        
+        /** Creates a csv record of the student's scores and answers, and the maximal scores and correct answers.
+         * 
+         * @returns {string} csv of student's results.
+         */
+        create_results_csv :  function () {
+            //todo: if questions are randomised at all, contain a line mapping to original ordering
+            const randomised = Numbas.results.examIsRandomised();
+            const questionMapping = Numbas.results.questionMapping();
+            let exam = Numbas.exam;
+
+            let header = [''];
+            let expectedAnswers = ['Ideal'];
+            let studentAnswers = ['Student'];
+            let trueOrder = ['True Order'];
+
+            header.push('Total Score');
+            expectedAnswers.push(exam.mark);
+            studentAnswers.push(exam.score);
+            trueOrder.push('');
+
+            for (let questionKey of Object.keys(exam.questionList)) {
+
+                let questionObject = exam.questionList[questionKey];
+                let questionName = questionObject.name||("Question "+questionKey);
+                let trueQuestionNumber = "Question " + questionMapping[questionKey]
+                header.push(questionName);
+                trueOrder.push(trueQuestionNumber);
+                expectedAnswers.push(questionObject.marks);
+                studentAnswers.push(questionObject.score);
+
+                for (let partKey of Object.keys(questionObject.parts)) {
+                    let partObject = questionObject.parts[partKey];
+                    let partName = partObject.name||("part "+partKey);
+                    header.push(questionName + " "+ partName + " Marks");

```

Comment:
For me, this looks incredibly ambiguous - it certainly wouldn't work for automated header-based merging of a bunch of csvs, unless we're assuming an absolute column listing ordering (which we can't, due to the randomisation), or relying on the 'machine-readable row' as the unique identifier. In which case, this kind of layout is simply for ease of putting into a spreadsheet software? (And again, that would fall apart with randomisation - I think if there's a random order, we might *need* to use/have a combiner webpage, which could do the nice formatting...)

Even so, I feel quite uncomfortable about a column existing whose person-readable header is " '' '' 'Answer'", and simply implying the question/part from its position in the table - in my PhD one of the file formats relied on item positioning to convey information and whilst it looked quite sleek, it started to really become an issue as more features came in, and we wanted to optimise processing, etc, so my instinct is to avoid that as far as possible!

Reply:
OK, then keep the three rows but concatenate all the things so you know what the prefixes are:

```
Question 1,
Question 1 a),
Question 1 a) Marks,
```

Reply:
I think I understand - if, say, it was 'The Question A' and 'Part 1' we could misread the addition as 'The Question' + 'A Part 1'? I'll think on this a little more, I'm not quite happy with the cascade but perhaps it makes sense.

Reply:
yep, that's exactly it: question names and part names can be anything you like, so there's no generic way of splitting up `question.name + ' ' + part.name`.